fuel-vm/src/interpreter/alu/wideint.rs

```diff
@@ -0,0 +1,365 @@
+
+                    write_bytes(&mut self.memory, owner_regs, dest_addr, wrapped.to_be_bytes())?;
+
+                    inc_pc(pc)
+                }
+
+                pub(crate) fn [<alu_wideint_mul_ $t:lower>](
+                    &mut self,
+                    dest_addr: Word,
+                    b: Word,
+                    c: Word,
+                    args: MulArgs,
+                ) -> Result<(), RuntimeError> {
+                    let owner_regs = self.ownership_registers();
+                    let (SystemRegisters { flag, mut of, mut err, pc, .. }, _) = split_registers(&mut self.registers);
+
+                    // LHS is only indirect if the flag is set
+                    let lhs: $t = if args.indirect_lhs {
+                        $t::from_be_bytes(read_bytes(&self.memory, b)?)
+                    } else {
+                        b.into()
+                    };
+                    // RHS is only indirect if the flag is set
+                    let rhs: $t = if args.indirect_rhs {
+                        $t::from_be_bytes(read_bytes(&self.memory, c)?)
+                    } else {
+                        c.into()
+                    };
+
+                    let (wrapped, overflow) = $t::overflowing_mul(lhs, rhs);
+
+                    if overflow && !is_wrapping(flag.into()) {
+                        return Err(PanicReason::ArithmeticOverflow.into());
+                    }
+
+                    *of = overflow as Word;
+                    *err = 0;
+
+                    write_bytes(&mut self.memory, owner_regs, dest_addr, wrapped.to_be_bytes())?;
+
+                    inc_pc(pc)
+                }
+
+                pub(crate) fn [<alu_wideint_div_ $t:lower>](
+                    &mut self,
+                    dest_addr: Word,
+                    b: Word,
+                    c: Word,
+                    args: DivArgs,
+                ) -> Result<(), RuntimeError> {
+                    let owner_regs = self.ownership_registers();
+                    let (SystemRegisters { flag, mut of, mut err, pc, .. }, _) = split_registers(&mut self.registers);
+
+                    // LHS is always indirect
+                    let lhs: $t = $t::from_be_bytes(read_bytes(&self.memory, b)?);
+
+                    // RHS is only indirect if the flag is set
+                    let rhs: $t = if args.indirect_rhs {
+                        $t::from_be_bytes(read_bytes(&self.memory, c)?)
+                    } else {
+                        c.into()
+                    };
+
+                    let lhs = [<to_prim_ $t:lower>](lhs);
+                    let rhs = [<to_prim_ $t:lower>](rhs);
+
+                    let result = match lhs.checked_div(rhs) {
+                        Some(d) => [<from_prim_ $t:lower>](d),
+                        None => {
+                            if is_unsafe_math(flag.into()) {
+                                $t::default() // Zero
+                            } else {
```

Comment:
if `is_wrapping == true`, should this behave like `overflowing_div` and set `*of = 1` without an error?

Reply:
Dividing by zero is an arithmetic error and not an overflow.  For instance the Rust function [`u64::overflowing_div`](https://doc.rust-lang.org/std/primitive.u64.html#method.overflowing_div) does the following:

> Returns a tuple of the divisor along with a boolean indicating whether an arithmetic overflow would occur. Note that for unsigned integers overflow never occurs, so the second value is always false.
> ### Panics
> This function will panic if rhs is 0.